cirq/devices/noise_model.py

```diff
@@ -58,20 +68,19 @@ def noisy_moments(self, moments: 'Iterable[cirq.Moment]',
             A sequence of OP_TREEs, with the k'th tree corresponding to the
             noisy operations for the k'th moment.
         """
-        if not hasattr(self.noisy_moment, '_not_overridden'):
-            result = []
-            for moment in moments:
-                result.append(self.noisy_moment(moment, system_qubits))
-            return result
 
-        if not hasattr(self.noisy_operation, '_not_overridden'):
-            result = []
-            for moment in moments:
-                result.append([self.noisy_operation(op) for op in moment])
-            return result
+    def _noisy_moment_impl_moments(self, moment: 'cirq.Moment',
+                                   system_qubits: Sequence['cirq.Qid']
+                                  ) -> 'cirq.OP_TREE':
+        return self.noisy_moments([moment], system_qubits)
 
-        assert False, 'Should be unreachable.'
+    def _noisy_moment_impl_operation(self, moment: 'cirq.Moment',
+                                     system_qubits: Sequence['cirq.Qid']
+                                    ) -> 'cirq.OP_TREE':
+        return [self.noisy_operation(op) for op in moment]
 
+    @value.alternative('noisy_moments', _noisy_moment_impl_moments)
```

Comment:
Force keyword arguments on this

`@value.alternative(requires='noisy_moments', implementation=_noisy_moment_impl_moments)`

or similar

Reply:
Done.